datasource/ethereum/src/ethereum_adapter.rs

```diff
@@ -47,6 +47,80 @@ where
         }
     }
 
+    fn traces(
+        &self,
+        logger: &Logger,
+        from: u64,
+        to: u64,
+        addresses: Vec<H160>,
+    ) -> impl Future<Item = Vec<Trace>, Error = Error> {
+        let eth = self.clone();
+        let logger = logger.to_owned();
+
+        retry("trace_filter RPC call", &logger)
+            .no_limit()
+            .timeout_secs(60)
+            .run(move || {
+                let trace_filter: TraceFilter = match addresses.len() {
+                    0 => TraceFilterBuilder::default()
+                        .from_block(from.into())
+                        .to_block(to.into())
+                        .build(),
+                    _ => TraceFilterBuilder::default()
+                        .from_block(from.into())
+                        .to_block(to.into())
+                        .to_address(addresses.clone())
+                        .build(),
+                };
+
+                let logger = logger.clone();
+                let logger_1 = logger.clone();
+                eth.web3
+                    .trace()
+                    .filter(trace_filter)
+                    .map(move |traces| {
+                        if traces.len() > 0 {
+                            if to == from {
+                                debug!(logger, "Received {} traces for block {}", traces.len(), to);
+                            } else {
+                                debug!(
+                                    logger,
+                                    "Received {} traces for blocks [{}, {}]",
+                                    traces.len(),
+                                    from,
+                                    to
+                                );
+                            }
+                        }
+                        traces
+                    })
+                    .from_err::<EthereumContractCallError>()
+                    .from_err()
```

Comment:
I wonder if we need the double `from_err` here. Could be, worth double-checking?

Reply:
We need go from the web3 error -> graph-node error -> failure error here, since graph-node errors implement `Fail`. Let me know if you have other ideas on how to get there

Reply:
I don't, just wanted to check.